docs: Add DynamicConfiguration tech doc (#1233)
aave/aave-v4: docs/features/DynamicConfiguration.md

The Governor retains the ability to force-migrate individual positions to the latest configuration via `updateUserDynamicConfig`. This mechanism is the primary tool for proactively managing risk when market conditions change between user interactions.

## Relationship to the Hub/Spoke Architecture

> Dynamic Risk Configuration is a Spoke-level concern. The three parameters it versions (Collateral Factor (CF), Liquidation Bonus (LB), and Liquidation Fee (LF)) govern how a user's collateral contributes to their health factor and how liquidation economics are computed. The Hub is unaware of these parameters; it maintains interest rate accounting, liquidity caps, and deficit state only. Spokes apply dynamic configs independently: the same underlying asset registered on two different Spokes carries independent configuration histories and independent `dynamicConfigKey` counters.

Reviewer: maybe should keep it formal and replace "configs" here to "configurations"
```suggestion
Dynamic Risk Configuration is a Spoke-level concern. The three parameters it versions (Collateral Factor (CF), Liquidation Bonus (LB), and Liquidation Fee (LF)) govern how a user's collateral contributes to their health factor and how liquidation economics are computed. The Hub is unaware of these parameters; it maintains interest rate accounting, liquidity caps, and deficit state only. Spokes apply dynamic configurations independently: the same underlying asset registered on two different Spokes carries independent configuration histories and independent `dynamicConfigKey` counters.
```

# Dynamic Risk Configuration

## Summary

Dynamic Risk Configuration is a spoke-level versioning mechanism that isolates collateralization parameters into per-Reserve configuration entries, each identified by a sequentially incrementing `dynamicConfigKey`. When governance updates collateralization parameters for a Reserve, the Spoke typically appends a new configuration entry rather than replacing the existing one (though governance can also update an existing key in place via `updateDynamicReserveConfig`). User positions retain a snapshot of the `dynamicConfigKey` active at the time of their last risk-bearing action. Parameter updates therefore do not immediately affect open positions; existing positions continue to evaluate under their snapshotted configuration until the user performs a risk-increasing action, at which point the Spoke rebinds the snapshot to the latest key. If the rebinding leaves the position under-collateralized, the action reverts.

> The Governor retains the ability to force-migrate individual positions to the latest configuration via `updateUserDynamicConfig`. This mechanism is the primary tool for proactively managing risk when market conditions change between user interactions.

Reviewer: I think we still need something as Yan mentions to say this is only for emergency scenarios, and not typically used

## Relationship to the Hub/Spoke Architecture

> Dynamic Risk Configuration is a spoke-level concern. The three parameters it versions (Collateral Factor (CF), Liquidation Bonus (LB), and Liquidation Fee (LF)) govern how a user's collateral contributes to their health factor and how liquidation economics are computed. The Hub is unaware of these parameters; it maintains interest rate accounting, liquidity caps, and deficit state only. Spokes apply dynamic configurations independently: the same underlying asset registered on two different Spokes carries independent configuration histories and independent `dynamicConfigKey` counters.

Reviewer: ```suggestion
Dynamic Risk Configuration is a spoke-level concern. The three parameters it encapsulates (Collateral Factor (CF), Liquidation Bonus (LB), and Liquidation Fee (LF)) govern how a user's collateral contributes to their health factor and how liquidation economics are computed. The Hub is unaware of these parameters; it maintains interest rate accounting, liquidity caps, and deficit state only. Spokes apply dynamic configurations independently: the same underlying asset registered on two different Spokes carries independent configuration histories and independent `dynamicConfigKey` counters.
```

The implementation spans three contracts in `src/spoke/`:

- `Spoke.sol`: exposes `addDynamicReserveConfig`, `updateDynamicReserveConfig`, and `updateUserDynamicConfig`, and implements all internal snapshot refresh logic.
- `SpokeStorage.sol`: declares the `_dynamicConfig` mapping and stores `_reserves` / `_userPositions` (whose `dynamicConfigKey` fields are defined in `ISpoke`).

> - `SpokeConfigurator.sol`: provides access-controlled convenience functions for common single-parameter updates that delegate to the above entry points on the target Spoke.

Reviewer: Hmm, SpokeConfigurator actually does have `updateDynamicReserveConfig` and `addDynamicReserveConfig` which are both not single parameter...

The maximum permitted `dynamicConfigKey` is `type(uint32).max` (approximately 4.29 billion entries per Reserve).

## Configuration Lifecycle

Configuration entries are created by appending a new key or updated by modifying an existing one. Both operations apply the same structural validation before committing state.

**Adding new configurations**

`addDynamicReserveConfig` creates a new configuration entry for a Reserve. The call increments the Reserve's `dynamicConfigKey` by one, stores the provided `DynamicReserveConfig` at that new key, advances the Reserve's `dynamicConfigKey` field to reference the new entry, and emits `AddDynamicReserveConfig`. New position snapshots created after this call bind to the new key.

Before storing, the Spoke validates three combined constraints under `InvalidCollateralFactorAndMaxLiquidationBonus`:

- `collateralFactor` must be strictly less than `100_00` BPS.

> - `maxLiquidationBonus` must be at least `100_00` BPS.

Reviewer 2: ```suggestion
- `maxLiquidationBonus` must be greater than or equal to `100_00` BPS.
```
to be consistent with line above

Updating an uninitialized key reverts with `DynamicConfigKeyUninitialized`, detected by checking that the stored `maxLiquidationBonus` is nonzero (all valid stored configurations carry a `maxLiquidationBonus` of at least `100_00`).

> **Warning:** `updateDynamicReserveConfig` affects all open user positions currently snapshotted at the target key. Reducing `collateralFactor` on a historical key reduces effective collateral value for those positions and may bring them closer to the liquidation threshold. The Governor should model the impact on bound positions before executing such an update.

The `SpokeConfigurator` exposes convenience wrappers that read the current latest configuration, replace the specified field, and call the underlying add or update function:

- `addCollateralFactor` / `updateCollateralFactor`
- `addMaxLiquidationBonus` / `updateMaxLiquidationBonus`
- `addLiquidationFee` / `updateLiquidationFee`
- `addDynamicReserveConfig` / `updateDynamicReserveConfig` — full configuration passthrough

## User Position Snapshots

> Each `UserPosition` stores a `dynamicConfigKey` field recording the configuration key currently bound to that user/reserve position for collateral evaluation. During health factor computation, the Spoke resolves the collateral factor for each Reserve from `_dynamicConfig[reserveId][userPosition.dynamicConfigKey].collateralFactor`. A collateral factor of zero causes the Reserve to contribute no value to the health factor regardless of the supplied amount.

Reviewer: Fine as is imo

On success, a `RefreshAllUserDynamicConfig` event is emitted. If the rebinding leaves the position under-collateralized, the call reverts with `HealthFactorBelowThreshold`, unwinding all state changes from the transaction.

**Actions that refresh a single Reserve's snapshot**

`setUsingAsCollateral` when enabling a Reserve as collateral calls `_refreshDynamicConfig`, which sets `userPosition.dynamicConfigKey = reserve.dynamicConfigKey` for the single Reserve being enabled and emits a `RefreshSingleUserDynamicConfig` event.

**Actions that do not refresh snapshots**

- `supply`
- `repay`
- `updateUserRiskPremium`
- `liquidationCall`

These actions evaluate health and risk using the user's existing snapshot keys. For `supply` and `repay`, the user's risk exposure decreases or stays neutral, so rebinding to latest configurations is not required. For `liquidationCall`, the liquidator acts on the position as-is; the health factor check uses the borrower's current snapshots.

## Health Factor Guard

When any action triggers a full snapshot refresh, the Spoke applies the following sequence atomically within the same transaction, before committing final state:

> 1. Iterate over all collateral Reserves in the position and set `userPosition.dynamicConfigKey = reserve.dynamicConfigKey` for each.

Author: I'll keep this one, as being tech docs it's a simple way to explain the HF guard without going too deep into details

2. Recompute the health factor using the newly bound configurations.
3. If `healthFactor < HEALTH_FACTOR_LIQUIDATION_THRESHOLD`, revert with `HealthFactorBelowThreshold`.

This guard prevents a user from taking a risk-increasing action while pinned to a configuration that, after rebinding, would leave the position under-collateralized. The rebind is unconditional; but if the transaction reverts due to the health factor check, the user's position remains unchanged.

> A position that is healthy under its current snapshot keys but under-collateralized under the latest Reserve configuration is effectively blocked from further risk-increasing actions until the user either repays debt, adds collateral via `supply`, or the Governor adjusts the latest configuration to one under which the position remains solvent.

Reviewer: ```suggestion
A position that is healthy under its current snapshot keys, but under-collateralized under the latest Reserve configuration, is effectively blocked from further risk-increasing actions until the user either repays debt, adds collateral via `supply`, or the Governor adjusts the latest configuration to one under which the position remains solvent.
```

- **Hub-level accounting**: CF, LB, and LF are not visible to the Hub. The Hub operates on share math, drawn indices, and liquidity caps only.
- **Per-Reserve snapshot migration**: `updateUserDynamicConfig` updates all collateral Reserves simultaneously. Individual Reserve snapshot updates are not exposed as a separate call.
- **Collateral Risk**: The risk premium parameter (`collateralRisk`) is stored in `ReserveConfig`, not in `DynamicReserveConfig`. It governs the risk premium interest component and is updated independently via `updateReserveConfig`. Dynamic configuration updates do not alter `collateralRisk`.
- **Liquidation execution logic**: Dynamic configurations supply the parameters used in the liquidation flow. The liquidation execution itself is documented separately.
- **Interest rate strategy**: Utilization-based borrow rates are a Hub-level concern and are unaffected by `DynamicReserveConfig`.
- **Cross-Spoke configuration sharing**: Each Spoke maintains its own `_dynamicConfig` mapping and independent key counters. Two Spokes connected to the same Hub asset do not share configuration history.

## Key Differences from Aave V3

> **Single global configuration per asset**: In V3, each asset carries exactly one risk configuration record. A governance update to Loan to Value, Liquidation Threshold, or Liquidation Bonus takes effect immediately for every open position borrowing against that asset. There is no mechanism to stage or scope the change to new positions only.

Author: We made a change to the style guide based on external feedback, "Aave V#" to be used always

# Dynamic Risk Configuration

## Summary

> Dynamic Risk Configuration is a spoke-level versioning mechanism (where each version is a snapshot of collateralization parameters) that isolates those parameters into per-Reserve configuration entries, each identified by a sequentially incrementing `dynamicConfigKey`. When governance updates collateralization parameters for a Reserve, the Spoke typically appends a new configuration entry rather than replacing the existing one (though governance can also update an existing key in place via `updateDynamicReserveConfig`). User positions retain a snapshot of the `dynamicConfigKey` active at the time of their last risk-bearing action. Parameter updates therefore do not immediately affect open positions; existing positions continue to evaluate under their snapshotted configuration until the user performs a risk-increasing action, at which point the Spoke rebinds the snapshot to the latest key. If the rebinding leaves the position under-collateralized, the action reverts.

Reviewer: Ideally, this first sentence should be clear if the reader does not read the parentheses. "those parameters" is not referenced outside the parentheses, but not a big deal
